### PR TITLE
Require components to implement `update`.

### PR DESCRIPTION
**Pull request by joshaber**
Addresses https://github.com/nathansobo/etch/issues/14.

@nathansobo and I talked about this yesterday and decided this was a reasonable path forward.

### CODE REVIEW

**Comment by joshaber** [comment]
That said, as the changed spec shows, this *does* feel a little funny/surprising.

1. Should `update` be required? Or at least log a warning?
2. Does `update` take, more or less, the same form as the constructor of receiving properties and scheduling a render?

I don't think we need to address either of those points now, but they're things to consider as Etch evolves.

**Comment by nathansobo** [comment]
> Should update be required? Or at least log a warning?

When would we warn or throw an exception? Maybe when a `ComponentWidget` is first created? I'm definitely open to either to prevent confusion.

> Does update take, more or less, the same form as the constructor: receiving properties and scheduling a render?

It does, but I haven't thought about a way of unifying them that feels right to me. From the perspective of using a component, it seems the most straightforward to construct an object with properties and later call `update` with properties. We could require the constructor to not take arguments and instead also call `update` when initializing the object, but that feels awkward to me.

If you want to implement a superclass in the package to experiment with some different patterns, that seems like it could be a good way to experiment with reducing boilerplate. You could implement the constructor and the `update` method in terms of hooks defined on subclasses and prescribe more about how properties are dealt with.

If you want to be more React-like, your superclass constructor and `update` implementations could perform a basic comparison on a `props` hash on the component and call various hooks such as `shouldComponentUpdate`, `componentWillReceiveProps`, `componentWillUpdate`, `componentDidUpdate`, etc.

I left this stuff out of Etch because it didn't seem as fundamental as the basic idea of calling `render` and using the result to create or update a DOM element. That felt like the essence of the problem and the rest felt more like advice around how to organize code. That said, maybe the props management pattern is standard enough to bake in a superclass for those that want to use it.

What gives me pause is wondering where it stops. Should we also model an internal `state` object with `setState` etc? I'm much less enthusiastic about that because of the fact that state has been more complex than a simple hash of key value pairs in a lot of the code I've worked on in Atom. The essential state of the editor is a product of multiple different indices tracking markers, syntax highlights, etc. It could never be boiled down into a `state` hash.

Maybe management of properties is different though, since they're more obviously always treated as a hash.

**Comment by joshaber** [comment]
I'm leaning towards making `update` required.

Without this PR, not implementing `update` is confusing for the reasons I outline in #14.

With this PR, not implementing `update` is confusing because re-renders wouldn't appear to have any effect.

TL;DR: Not implementing `update` is potentially confusing in all cases.

>When would we warn or throw an exception? Maybe when a ComponentWidget is first created?

First creation seems good :+1: 

Thoughts? I can Do The Needful if that sounds like a good path forward to you.

**Comment by nathansobo** [comment]
@joshaber :+1: on making it required. You can always intentionally do nothing. If you subclass from some component that implements it for you, even easier.

**Comment by joshaber** [comment]
:+1: Changed it to throw on initialize if there is no `update`.

**Comment by nathansobo** [comment]
Hdatkn!

**Comment by nathansobo** [comment]
Woah, my fingers won't on the home row: Thanks!

**Comment by nathansobo** [comment]
I added you as an owner of the module so you can publish.

**Comment by joshaber** [comment]
:heart:
